### ISSUE TITLE
Fundamental improvements to how database services set custom credentials and related commands

### ISSUE THREAD

**Issue by RobinHoutevelts** [lando/lando#853]
**Tell us about your `.lando.yml`**

```yml
name: mysite
recipe: drupal8
services:
  database:
    creds:
      user: drupal
      password: drupal
      database: drupal
```

**Tell us about the command you were running**

```bash
lando init
lando start
lando db-import db.sql.gz
```

**Tell us about the error you got**
```
$ lando db-import db.sql.gz          
Preparing to import /app/db.sql.gz into drupal8 on database:3306 as drupal8...
ERROR 1045 (28000): Access denied for user 'drupal8'@'mysite_appserver_1.mysite_default' (using password: YES)
Destroying all current tables in drupal8... 
NOTE: See the --no-wipe flag to avoid this step!
Gunzipped file detected!
Importing /app/db.sql.gz...
ERROR 1045 (28000): Access denied for user 'drupal8'@'mysite_appserver_1.mysite_default' (using password: YES)
  96KiB 0:00:00 [16.1MiB/s] [>                                 ]  3%            

Import failed.
```

**Tell us generally about your bug**

I tried to import a database but got a `ERROR 1045 (28000): Access denied for user 'drupal8'@...` error.

While I can define `services.database.creds` in my `.lando.yml` file, it's values are not used by `db-import`.

Nor are those values put in the `appserver` `DB_*` environment variables.

**Tell us more**

I'm having trouble understanding how to override the database credentials that a recipe creates ( you can't , well not entirely).

People end up creating a `.lando.yml` file like I did. I have no idea where I came up with this, nor why I expected it to work, but it appears I'm not the only one (https://github.com/lando/lando/issues/843, https://github.com/lando/lando/issues/671, https://github.com/lando/lando/issues/593)

**Comment by pirog** [comment]
@RobinHoutevelts Great write up here and even better explanation of what is going on. I think you correctly identified some other issues that this likely impacts (and im going to add another)

#843 #671 #853 #593 + #803 #727 

Reading through all these issues i think we should increase the scope of this ticket to address the following concerns

### 1. Poor Documentation

It is not clear at all what a user must do to change the database credentials once they've been set. We need to make sure this is documented in some sort of warning in all relevant services (and recipes depending on what we choose below). 

This is probably the most straightforward change we need to make and a no brainer. 

### 2. How do recipes/services interact wrt db (or any services) creds? 

You are currently not able to set db creds in a recipes config section. However, intrepid users have used landos services layer to override recipe config. This works insofar as you can change the db creds on a destroy/start but the change does not impact related config in attached services: most notably appserver envvars which are currently used for application connection config and things like db-import commands. There are a few ways we could make this work better and its possible we could implement many:

1. Do an implementation that accomplishes what #854 does, except im guessing we would want to do this further down the stream to ensure the stream flows one way eg: recipe config is unpacked into services at the recipe layer and then passed down into the services layer to be mixed in with any other services. 
2. We could just not allow recipe db creds to be overriden 
3. We could add creds as part of the recipe config
4. We could add something like what is proposed #727 
5. We could consider whether our current model of `DB_*` set in the appserver makes sense

### 3. How should `db-import/export` work given you could have multiple databases of different types

Right now we are running `db-import` and `db-export` from the `appserver` which means it falls prey to the same as above. This coupling made sense early on but we are starting to run into limitations (again the above) and also potential client/server mismatches as @serundeputy points out in #803. We should probably be running these commands directly from the target service and injecting the needed ENVVARS on each command run instead of piggybacking off the values of the `appserver`. 

1. Need to figure out how to dynamically set the tooling `service` based on the target database (eg the current host option for those commands)
2. We need to figure out how to inject the options into the script (it might already do this?)

**Comment by RobinHoutevelts** [comment]
## TL;DR

- I 100% agree ENVVARS should be deprecated. Users should provide their own `.env` based on their recipe or overrides.
- I 100% agree the `db-*` commands shouldn't target the appserver.
- I'm not sold on shrinking the config options down to _only_ the target service. We'd be unnecessarily breaking a lot of existing scripts.
- I think a recipe's `build()` method should accept a `services` var containing user-defined services<sup>(possible overrides)</sup> in `.lando.yml`
  - This might be out of scope for this ticket?

----

> How do recipes/services interact wrt db (or any services) creds?

> Do an implementation that accomplishes what #854 does, except im guessing we would want to do this further down the stream to ensure the stream flows one way eg: recipe config is unpacked into services at the recipe layer and then passed down into the services layer to be mixed in with any other services.

To me it feels like this is what we have to aim for.

This is my take on this:

The recipe's `build()` method should accept a `services` var that contains the pre-defined services made in `.lando.yml`.

Much like how most recipes start off with a `// Start by cheating` and calling another recipe, we now would be  starting with whatever the user pre-defined.

```diff
diff --git a/plugins/lando-recipes/recipes/lamp/lamp.js b/plugins/lando-recipes/recipes/lamp/lamp.js
index a0d4e998..ba1caa83 100644
--- a/plugins/lando-recipes/recipes/lamp/lamp.js
+++ b/plugins/lando-recipes/recipes/lamp/lamp.js
@@ -114,14 +114,14 @@ module.exports = function(lando) {
   /*
    * Helper to return services config
    */
-  var services = function(config) {
+  var buildServices = function(config, services) {
 
     // Get some options
     var phpVersion = _.get(config, 'php', '7.1');
     var database = _.get(config, 'database', 'mysql');
 
     // Build our default set of services
-    var services = {
+    services = merger({
       appserver: {
         type: 'php:' + phpVersion,
         via: 'apache',
@@ -137,7 +137,7 @@ module.exports = function(lando) {
           database: config._recipe
         }
       }
-    };
+    }, services);
 
     // Mix in any additional config
     if (_.has(config, 'webroot')) {
@@ -332,14 +332,14 @@ module.exports = function(lando) {
   /*
    * Build out LAMP
    */
-  var build = function(name, config) {
+  var build = function(name, config, services) {
 
     // Start up our build
     var build = {};
 
     // Get our things
     build.proxy = proxy(name);
-    build.services = services(config);
+    build.services = buildServices(config, services);
     build.tooling = tooling(config);
 
     // Return the things

```

-- It should be fairly easy to implement and it's backwards compatible.

> We could add something like what is proposed #727

#727 is something that would definitely be handy!

> We could consider whether our current model of DB_* set in the appserver makes sense

It doesn't and is _very_ susceptible to bugs. For example:

I have a project with a `.env` where `DB_HOST` points to a remote db on AWS.
While my `creds` are correct, lando won't override my `.env` ( thank god ). But that also means when I run `db-import` it will try to import to AWS instead of the `database` service 😢 

>  We should probably be running these commands directly from the target service and injecting the needed ENVVARS on each command run instead of piggybacking off the values of the appserver.

Yes! I'm not 100% sold on shrinking the config options down to _only_ the target service. We can stay backwards compatible by first loading the creds by calling the info method of the target service and then override those values with whatever the user passed by.

If not we'd be breaking a lot of people's scripts.

**Comment by pirog** [comment]
@RobinHoutevelts i definitely dont think we want to be checking for user defined services in recipes. these are all mixed in further down the stream and as far as i can tell already do what you've proposed in the diff above. 

I think this ticket just happens to surface a few places where that flow breaks but i think its more a referendum on what we are trying to do with ENVVARS and db-import than the general build flow from recipes to services

Re: shrinking db-import options, i think we want to remain "quietly" backwards compatible here eg we aren't going to surface the old options anymore but our script will still know how to handle them. we are still technically beta so backwards compatibility is not something users should assume. that said we should def provide a "best effort" to make as few breaking changes as possible.

**Comment by dustinleblanc** [comment]
I am not clear why exposing env vars like DB_THANG is a bad thing. Don't users use these env vars if they aren't doing a Pantheon style thing? Perhaps just namespacing them is a better idea?

**Comment by pirog** [comment]
@dustinleblanc i think the main issues are

1. Unreliability in terms of downstream overrides, eg you cant change your db creds in a recipe with a service override and have those ENVVARS update without a pretty nasty break in how the build flow is supposed to work
2. Collision with other things given the common naming convention
3. Hardcoded vars and general unavailability eg we only set these ENVVARS on certain services and on certain recipes 

The good news is that implementing #727 basically solves all of the above and we can keep the current hardcoded `DB_*` envvars around as well and set a deprecation date with encouragement to use the new `$LANDO_INFO`

**Comment by dustinleblanc** [comment]
Ah right using LANDO_INFO is essentially exactly the 'PANTHEON' way of doing it (like PRESSFLOW_SETTINGS). Might be a bit of friction for some users, we might want to document a solid way to extract the env vars (similar to what Pantheon does in our docs).

The more I think of it though, most of the time I either just hardcode the creds in the settings files for apps that don't use env injection or just edit the .env file on apps that do, I am worrying about it too much.

**Comment by pirog** [comment]
@dustinleblanc, yah i think we basically want to change all the recipe/services that mention injected ENVVARS with a deprecation notice for the `DB_*`ish ones, a reference to the new `LANDO_INFO` and then a link to a tutorial with some code snippets so users can see how to unpack that in various langz

**Comment by RobinHoutevelts** [comment]
> @RobinHoutevelts i definitely dont think we want to be checking for user defined services in recipes. these are all mixed in further down the stream and as far as i can tell already do what you've proposed in the diff above.

That's too late if you want a recipe to be able to handle service overrides.
But perhaps you're right. If someone needs to override services provided by a recipe, it just means that recipe needed more options in their config.

Anyway, it's definitely out of scope for this ticket 👍

**Comment by pirog** [comment]
@RobinHoutevelts recipes ALREADY handle service overrides, this is one of the core features of Lando.
